l10n: zh_CN: Add translations for Git glossary

jiangxin opened the pull request:
Add translations of Git glossary (most of them are from the command
`git help glossary`) in the header of `zh_CN.po`.  Also fixes some translations
according to this glossary, such as "pathspec", "repository", "refspec".

Signed-off-by: Jiang Xin worldhello.net@gmail.com

jiangxin committed:
l10n: zh_CN: Add translations for Git glossary

Add translations of Git glossary (most of them are from the command `git
help glossary`) in the header of `zh_CN.po`.  Also fixes some translations
according to this glossary, such as "pathspec", "repository", "refspec".

Signed-off-by: Jiang Xin <worldhello.net@gmail.com>
Signed-off-by: Ray Chen <oldsharp@gmail.com>

jiangxin commented:
Please help to review, @oldsharp.

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   hunk                             |  补丁片段
+#   index                            |  索引
+#   index entry                      |  索引条目
+#   master                           |  master（缺省分支名）
```
The word "default" is translated as "默认" in the body text of this file.  Should we unify this translation?  Same issue in line 62.

jiangxin replied:
s/缺省/默认/gc

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   pack index                       |  包索引
+#   pathspec                         |  路径规格
+#   pattern                          |  模式
+#   parent                           |  parent
```
Consider translating "parent" as "父提交".

jiangxin replied:
I wonder if we could find some politically correct translations. e.g.

```
parent element: 母体元素;母元素
parent firm: 母公司
parent form: 父窗体；母型；上层窗体
parent language: 母语；原始语；源语
```

oldsharp replied:
Yes.  Things are little bit tricky here.

Beside the choosing of "父/母", here's another related issue which comes into my mind:

The manual of `git-commit-tree` told us:

> A commit object may have any number of parents.  With exactly one parent, it is an ordinary commit.  Having more than one parent makes the commit a merge between several lines of history.  Initial (root) commits have no parents.

When a commit has more than 3 logical (directly) predecessors, the term "parent" itself is not accurate - none of us has more than **two** parents!
In such a case, the translation work is somehow embarrassing - "父提交", "母提交", and even "双亲提交" - none of them is 100% correct :-(
Maybe "前任提交" is a better choice.

However, we may consider this in another angle.
The Git manual use the term "parent" in some loose way - it just stands for all the object(s) in the logical predecessor(s) list.
When taking this into consideration, I think we can safely translate the term "parent" as "父提交" - a Git object may have many logical predecessors, each of which is a "父提交".

Return to the issue of choosing "父提交"/"母提交": I personally prefer "父提交" because I think  it has more inheritability with the term "ancestor".  Hopes the human rights organization would not argues against this ;-)

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   grafts                           |  （提交）嫁接
+#   hash                             |  哈希值
+#   head                             |  头、分支
+#   HEAD                             |  头指针（当前分支）
```
"HEAD" is left untranslated in many places of the body. Maybe we should fix (some places of) this.

jiangxin replied:
OK, I will leave "HEAD" untranslated here.

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   commit-ish (also committish)     |  提交号
+#   core Git                         |  核心 Git 工具
+#   DAG                              |  有向无环图
+#   dangling object                  |  摇摆对象
```
Its rare in computer science area, to translate "dangling" as "摇摆".  Maybe "悬空对象" is a better choice?  (See https://zh.wikipedia.org/wiki/%E8%BF%B7%E9%80%94%E6%8C%87%E9%92%88 )

jiangxin replied:
I prefer 摇摆. some dangling objects are useful, others are just trash.

oldsharp replied:
Oh I see.

The conception of Git's dangling object is unlike the wild pointer in some programming language - the Git objects may be referenced again later.

Yeah, that's my fault.  Agree with "摇摆".

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   plumbing                         |  管件（Git 底层核心命令的别称）
+#   porcelain                        |  瓷件（Git 上层封装命令的别称）
+#   prune                            |  清除
+#   pull                             |  拉
```
pull -> 拉，拉取

While "pull" translated as "拉" is simple and accurate, in some context "拉取" is a better choice (especially where "pull" appears as a noun, e.g., "拉取操作").  So we can consider providing both options and leave the decision to the translator.

jiangxin replied:
agree.

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   porcelain                        |  瓷件（Git 上层封装命令的别称）
+#   prune                            |  清除
+#   pull                             |  拉
+#   push                             |  推
```
push -> 推，推送

Same reason as the "pull" entry.

jiangxin replied:
agree.

oldsharp commented on po/zh_CN.po:
```diff
@@ -4575,7 +4667,7 @@ msgstr "参数 --include/--only 不跟路径没有意义。"
 #: builtin/commit.c:1195
 msgid "Clever... amending the last one with dirty index."
-msgstr "聪明... 在索引不干净下修补最后的提交。"
+msgstr "聪明... 用脏索引修补带最后一个提交。"
```
Could not understand the word "带" in this line, maybe we should remove it?

jiangxin replied:
Right, it's a typo.

oldsharp commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   reflog                           |  引用日志
+#   refspec                          |  引用规格
+#   remote-tracking branch           |  远程跟踪分支
+#   repository                       |  仓库（亦为版本库）
```
I suggest removing the note "（亦为版本库）" to avoid potential confusion in the future.

networm replied:
> bare repository                  |  纯仓库
> shallow repository               |  浅（克隆）仓库

所以我认为这里的“（亦为版本库）”容易引起歧义，建议删除。

jiangxin replied:
Fine.

networm commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   branch                           |  分支
+#   bundle                           |  包
+#   bypass                           |  绕过
+#   cache                            |  索引（的别称）
```
cache 下面都翻译为缓存，这里为什么指定为索引？我搜索了一下全文，未找到相关用法。

oldsharp replied:
@networm In Git context, when we talking about "cache", we mean the index area.  However, in other context, "cache" may stand for "缓存".

Maybe some of the outdated translation in the body should be updated.

networm commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   shallow repository               |  浅（克隆）仓库
+#   signed tag                       |  签名标签
+#   stage                            |  暂存（即索引）
+#   stash                            |  进度保存
```
正文内容中有的地方翻译为“进度保存”，有的翻译为“保存进度”，有的翻译为“进度”，未统一。
感觉“保存进度”更合乎语法。

oldsharp replied:
@networm In my opinion: stash translated as "进度保存" when as a noun, while translated as "保存进度" when as a verb.

networm commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   tree-ish (also treeish)          |  树对象（或可以解析为一树对象）
+#   unmerged index                   |  未合并索引
+#   unreachable object               |  不可达对象
+#   unstage                          |  撤出暂存区（即撤出索引）
```
> stage -> 暂存（即索引）
> unstage -> 撤出暂存区（即撤出索引）

unstage 我认为应该使用与 stage 相同的动词，即“取消暂存”。

jiangxin replied:
stage -> n. 暂存区（即索引）; v. 暂存

networm commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#
+#   English                          |  Chinese
+#   ---------------------------------+--------------------------------------
+#   3-way merge                      |  三路合并
```
3-way merge 翻译为“三方合并”似乎更常见、更合语境，因为这里并不是真的三条线合成一条线，而是两条线合成一条线，同时考虑共同的父结点。

networm replied:
Merge (revision control) - Wikipedia, the free encyclopedia
https://en.wikipedia.org/wiki/Merge_(revision_control)

参见 Three-way merge 小节。

jiangxin replied:
s/三路合并/三方合并/g

networm commented on po/zh_CN.po:
```diff
@@ -8,12 +8,104 @@
+#   parent                           |  parent
+#   pickaxe                          |  pickaxe
+#   plumbing                         |  管件（Git 底层核心命令的别称）
+#   porcelain                        |  瓷件（Git 上层封装命令的别称）
```
瓷件 并未找到维基百科或百度百科对应的词条，并不知道这是什么意思。

oldsharp replied:
@networm I think the translation of "plumbing" and "porcelain" is nice, since "管件" and "瓷件" are appropriate and vivid for this two words - especially when we take the original meaning of them into consideration.

After all, the terms "plumbing" command and "porcelain" command themselves are created by the Git community :-)

networm replied:
OK, I think it's okay.

oldsharp commented:
@jiangxin My review work is done.  Please kindly check the line notes I added during the review process, and tell me what's your opinion.  Thanks.

Additionally, here I'd like to provide an extra list of Git terminology for our glossary - they may not included in the official Git glossary, but may also be helpful with the future translation work:
- abbreviate -> 简写（的 SHA-1 值）
- amend -> 修补
- ancestor -> 祖先，祖先提交
- checksum -> 校验，校验和
- clone -> 克隆
- conflict -> 冲突
- dumb HTTP -> 哑 HTTP 协议
- fork -> 派生，复刻
- header -> 头信息
- log -> 日志
- loose object -> 松散对象
- loose refs -> 松散引用
- packfile -> 包文件
- patch -> 补丁
- remote -> 远程，远程仓库
- replay -> 重放
- repo -> 仓库
- revert -> 还原
- smart HTTP -> 智能 HTTP 协议
- squash -> 压缩
- submodule -> 子模组
- tagger -> 打标签者
- track -> 跟踪
- unpack -> 解包
- upstream -> 上游

Please also check the list above, and consider adding these entries into our glossary.  Thanks.

What if we still missed something?  That's fine - our glossary header is tend to be a dynamically growing list.  Any reasonable appending contribution request from the community is welcomed.  So I personally do not worry about that ;-)

jiangxin committed:
l10n: TEAMS: stash inactive zh_CN team members

Add Ray Chen as member of zh_CN l10n team member, and move other
inactive zh_CN l10n team members to the header of zh_CN.po.

Signed-off-by: Jiang Xin <worldhello.net@gmail.com>

oldsharp commented on po/zh_CN.po:
```diff
@@ -4,16 +4,137 @@
+#   unmerged index                   |  未合并索引
+#   unpack                           |  解包
+#   unreachable object               |  不可达对象
+#   unstage                          |  取消暂存
```
Just remind: the body of the file is not fully updated with the new translation version of "unstage".  However, we can open another PR to fix this later.

oldsharp commented on po/zh_CN.po:
```diff
@@ -4,16 +4,137 @@
+#   fast-forward                     |  快进
+#   fetch                            |  获取
+#   file system                      |  文件系统
+#   fork                             |  派生，复刻
```
https://github.com/jiangxin/git/blob/zh_CN/git-glossary/po/zh_CN.po#L7444 : 

Original text: `msgid "find where <commit> forked from reflog of <ref>"`
Old version translation: `msgstr "查找 <提交> 从<引用>日志的何处分支的"`
Update suggestion: `msgstr "根据引用日志 <引用> 查找 <提交> 的派生处"`

Refer to the `--fork-point` option in the `git-merge-base` manual.

networm replied:
复刻 这个翻译并没有体现出继承这个关系，所以建议去除，只使用派生。

oldsharp replied:
@networm Agreed.  Will remove it later.

oldsharp commented on po/zh_CN.po:
```diff
@@ -4,16 +4,137 @@
+#   patch                            |  补丁
+#   pathspec                         |  路径规格
+#   pattern                          |  模式
+#   pickaxe                          |  pickaxe
```
Maybe we can figure out a proper translation of "pickaxe".  Refer to the "DIFFCORE-PICKAXE" section in the `gitdiffcore` manual.

louy2 replied:
Propose: pickaxe -> 挖掘

oldsharp replied:
@louy2 Good suggestion, would you like to open a PR for it?

oldsharp commented:
@jiangxin The 2nd round of the review process has been done. Please have a check.  Thanks.

oldsharp commented:
cherry-picked.

oldsharp commented:
Note that we left 2 issues unresolved yet:
- the "cache" translation (see https://github.com/jiangxin/git/pull/6#discussion_r36087962)
- the "pickaxe" translation (see https://github.com/jiangxin/git/pull/6#discussion_r36154708)

Consider resolve them in the future.
